tests/src/common/lease.rs

```diff
@@ -341,79 +341,103 @@ fn send_open_ica_response<'r>(
+    remote_tx_count: usize,
+) {
+    response.expect_submit_tx(connection_id, "0", remote_tx_count);
 
-fn check_state_opened(app: &mut App, lease: &Addr) {
-    if !matches!(fetch_state(app, lease), StateResponse::Opened { .. }) {
-        panic!("Opening lease failed! Lease is not yet it opened state!");
-    }
+    response.unwrap_response()
 }
 
-fn send_swap_response<DownpaymentC, LeaseC>(
+fn do_swap<DownpaymentC, Lpn, LeaseC>(
     app: &mut App,
-    lease: &Addr,
-    swap_out: Coin<LeaseC>,
-    expected_amounts_count: usize,
+    lease_addr: Addr,
+    ica_addr: &str,
+    (downpayment, exp_borrow, exp_lease): (Coin<DownpaymentC>, Coin<Lpn>, Coin<LeaseC>),
 ) where
     DownpaymentC: Currency,
+    Lpn: Currency,
     LeaseC: Currency,
 {
-    let amounts_out: Vec<Amount> = if currency::equal::<DownpaymentC, LeaseC>() {
-        vec![swap_out.into()]
+    let downpayment_equals_lease: bool = currency::equal::<DownpaymentC, LeaseC>();
+
+    let exp_swap_out: Coin<LeaseC> = if downpayment_equals_lease {
+        exp_lease - price::total(downpayment, Price::identity())
     } else {
-        let downpayment_out = 1;
-        let borrow_amount = Into::<Amount>::into(swap_out) - downpayment_out;
-        vec![downpayment_out, borrow_amount]
+        exp_lease
     };
 
-    assert_eq!(amounts_out.len(), expected_amounts_count);
+    app.send_tokens(
+        Addr::unchecked(ica_addr),
+        Addr::unchecked(ADMIN),
+        &[cwcoin(exp_borrow), cwcoin(downpayment)][..1 + usize::from(!downpayment_equals_lease)],
+    )
+    .unwrap();
 
-    check_state_opening(app, lease);
+    assert_eq!(
+        app.query().query_all_balances(ica_addr).unwrap().as_slice(),
+        &[cwcoin(downpayment)][..usize::from(downpayment_equals_lease)],
+    );
 
-    let swap_resp = swap_exact_in_resp(amounts_out);
+    app.send_tokens(
+        Addr::unchecked(ADMIN),
+        Addr::unchecked(ica_addr),
+        &[cwcoin(exp_swap_out)],
+    )
+    .unwrap();
 
-    () = send_response(app, lease, swap_resp)
-        .ignore_response()
-        .unwrap_response();
+    send_swap_response(
+        app,
+        lease_addr.clone(),
+        exp_swap_out,
+        downpayment_equals_lease,
+    );
+
+    check_state_opened(app, lease_addr);
 }
 
-fn swap_exact_in_resp<I>(amounts: I) -> Binary
-where
-    I: IntoIterator<Item = Amount>,
+fn send_swap_response<LeaseC>(
+    app: &mut App,
+    lease: Addr,
+    swap_out: Coin<LeaseC>,
+    downpayment_equals_lease: bool,
+) where
+    LeaseC: Currency,
 {
-    let msgs = amounts
-        .into_iter()
-        .map(swap_trx::build_exact_amount_in_resp);
-    trx::encode_msg_responses(msgs).into()
+    let downpayment_equals_lease_as_amount: Amount = (!downpayment_equals_lease).into();
+
+    let amounts_out = [
+        Amount::from(swap_out) - downpayment_equals_lease_as_amount,
+        downpayment_equals_lease_as_amount,
+    ]
+    .into_iter()
+    .filter(|&amount: &Amount| amount != 0);
```

Comment:
too cryptic ... why do we mix bool with Amount?

Reply:
My intent was to avoid adding more logic depending on whether down-payment and lease are equal. I followed previous code's approach and wanted to skip any steps that would complicate it.